presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java

```diff
@@ -0,0 +1,453 @@
+import static java.sql.Types.LONGVARCHAR;
+import static java.sql.Types.NVARCHAR;
+import static java.sql.Types.TIMESTAMP;
+import static java.sql.Types.TIMESTAMP_WITH_TIMEZONE;
+import static java.sql.Types.TIME_WITH_TIMEZONE;
+import static java.sql.Types.VARCHAR;
+import static java.util.Collections.nCopies;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+import static org.apache.phoenix.coprocessor.BaseScannerRegionObserver.SKIP_REGION_BOUNDARY_CHECK;
+import static org.apache.phoenix.query.QueryConstants.NULL_SCHEMA_NAME;
+import static org.apache.phoenix.util.SchemaUtil.ESCAPE_CHARACTER;
+
+public class PhoenixClient
+        extends BaseJdbcClient
+{
+    private final Configuration configuration;
+
+    @Inject
+    public PhoenixClient(PhoenixConfig config)
+            throws SQLException
+    {
+        super(
+                new BaseJdbcConfig(),
+                ESCAPE_CHARACTER,
+                new DriverConnectionFactory(
+                        DriverManager.getDriver(config.getConnectionUrl()),
+                        config.getConnectionUrl(),
+                        Optional.empty(),
+                        Optional.empty(),
+                        config.getConnectionProperties()));
+        requireNonNull(config, "config is null");
+        configuration = new Configuration(false);
+        ConnectionInfo connectionInfo = PhoenixEmbeddedDriver.ConnectionInfo.create(config.getConnectionUrl());
+        connectionInfo.asProps()
+                .forEach(prop -> configuration.set(prop.getKey(), prop.getValue()));
+        config.getConnectionProperties()
+                .forEach((k, v) -> configuration.set((String) k, (String) v));
+    }
+
+    public PhoenixConnection getConnection(JdbcIdentity identity)
+            throws SQLException
+    {
+        Connection connection = connectionFactory.openConnection(identity);
+        try {
+            return connection.unwrap(PhoenixConnection.class);
+        }
+        catch (Exception e) {
+            connection.close();
+            throw new PrestoException(PHOENIX_QUERY_ERROR, "Couldn't open Phoenix connection", e);
+        }
+    }
+
+    public org.apache.hadoop.hbase.client.Connection getHConnection()
+            throws IOException
+    {
+        return HBaseFactoryProvider.getHConnectionFactory().createConnection(configuration);
+    }
+
+    public void execute(ConnectorSession session, String statement)
+    {
+        try (Connection connection = connectionFactory.openConnection(JdbcIdentity.from(session))) {
+            execute(connection, statement);
+        }
+        catch (SQLException e) {
+            throw new PrestoException(PHOENIX_QUERY_ERROR, "Error while executing statement", e);
+        }
+    }
+
+    @Override
+    protected Collection<String> listSchemas(Connection connection)
+    {
+        try (ResultSet resultSet = connection.getMetaData().getSchemas()) {
```

Comment:
I think this method can be simplified to
```java
return ImmutableSet.builder()
        .add(DEFAULT_SCHEMA)
        .addAll(super.listSchemas(connection))
        .build();
```
(this would return `JDBC_ERROR` on error but that seems fine)

Reply:
There is another subtle difference from the parent class implementation here - we check if `schemaName` is `null`.  Phoenix can return `null` in `connection.getMetaData.getSchemas`, so without this check we'll get a NPE.